architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,641 @@
+
+For anyone skeptical about this, the concept of resource provider is already present in **every** resource id/URL. This is the meaning of the `providers` segment. Currently URLs like `/planes/radius/local/providers` don't have any functionality associated with them, and return a 404 - this proposal addresses this gap.
+
+One of these lessons-learned is the ability to register multiple regions/locations where a resource provider can operate. In particular users will need the ability to gradually roll out functionality across regions rather than updating all regions at once. This is complex, and doesn't help us in the short term. The approach taken here is aligned with our current strategy for region/location in Radius. The APIs are designed for multi-region but the implementation assumes a single region is used.
+
+#### Resource Types
+
+The registration APIs add a few new resource types:
+
+- `System.Resources/resourceProviders` for resource providers.
+- `System.Resources/resourceProviders/resourceTypes` for resource types.
+- `System.Resources/resourceProviders/resourceTypes/apiVersions` for API versions of resource types.
+- `System.Resources/resourceProviders/locations` for the configuration of a resource provider in a specific location/region.
+
+These new resource types define the full CRUDL set of APIs. These APIs are optimized for operations that need to modify or deeply understand resource types. 
+
+These are our first nested resource type in UCP/Radius. This is a good usecase for a nested resource types because:
+
+- The `apiVersions` body will be be large due to the inclusion of an OpenAPI schema. 
+- Users need granular registration and unregistration of individual resource types and API versions. (See scenario 2 above).
+- Modeling `location` as a resource optimizes for per-region rollouts and validation of new functionality.
+
+---
+
+These new resource types are scoped to the Radius plane. 
+
+> eg: `/planes/radius/local/providers/System.Resources/resourceProviders`
+
+The rationale for this is to support registration/unregistration per-tenant of Radius rather than per-installation.
+
+#### Query Patterns
+
+The nested resource pattern is the best approach to support the data access patterns needed to implement UDTs. 
+
+Some examples:
+
+- UCP Request Routing:
+  - UCP knows the location, resource type, and API version.
+  - UCP can query a single `locations` resource to check for support and find the downstream API address.
+- Dynamic-RP Validation:
+  - Dynamic-RP knows the resource type and API version.
+  - Dynamic-RP can query a single `apiVersions` resource to fetch its schema for validation. 
+
+#### Resource Type Summaries
+
+The proposal also defines two new **non-resource** APIs:
+
+- `GET /planes/radius/{plane}/providers/`: list the most commonly used fields all resource providers.
```

Comment:
```suggestion
- `GET /planes/radius/{plane}/providers/`: lists the most commonly used fields all resource providers.
```

Reply:
How will the system decide on what the most commonly used fields are? Are we talking about returning all the fields of all resource providers? What are the most commonly used fields?

Reply:
We will decide. We're really smart 😉 

This will be driven by scenarios/needs. For example, what does the `rad` CLI need to know to give a good experience? Consider that question times 3-4 other scenarios and you have the idea.